### PR TITLE
feat: add keys to inventory

### PR DESCRIPTION
**Pull request by wingbladedota**
## Description

This PR adds keys to the inventory, and adds methods to add keys to the inventory, and to query for keys in the inventory. Check https://zomsa.atlassian.net/jira/software/projects/APLIB/boards/1/backlog?selectedIssue=APLIB-71 for more info. The interaction between doors and keys is not yet added in this PR.

## Testability

Open keytestlevel. Add a test function to the inventory script that adds some keys and queries for them. An example would be to add KeyTest(); to the start, and then to copy paste this somewhere else in the inventory script:

    private void KeyTest()
    {
        Key testkey;
        Key testkey2;
        testkey = new Key(1);
        testkey2 = new Key(2);
        PickUpKey(testkey);
        PickUpKey(testkey2);
        Debug.Log(_keyRing);
//now you need to add a door with a certain ID, and do some queries

    }

## Checklist
- [ ] Set the proper pull request name
- [x] Merged main into your branch
- [x] Added a scene to the game for your changes

### CODE REVIEW

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]

File: aplib.net-demo/Assets/Scripts/Inventory/Key.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
+public class Key : Item
+{
+    /// <summary>
+    /// The key ID, to check which door it can open
+    /// </summary>
+    public int id;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Inventory/Key.cs:6]: Also note our identifier naming conventions.

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]

File: aplib.net-demo/Assets/Scripts/Inventory/Key.cs (hunk 1 above)
Comment (JensSteenmetz) [reply aplib.net-demo/Assets/Scripts/Inventory/Key.cs:6]: Dit zou een `public PascalCase; // ------------------------ // Public nonstatic fields` zijn.

File: aplib.net-demo/Assets/Scripts/Inventory/Inventory.cs [hunk 2]
```diff
@@ -49,20 +48,52 @@ public void PickUpItem(Item item, float uses = 1)
...
+    public bool KeyQuery(int doorId)
+    {
+        foreach (Key k in _keyRing)
+        {
+            if (k.id == doorId)
+            {
+                _ = _keyRing.Remove(k);
+                return true;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (JensSteenmetz) [reply aplib.net-demo/Assets/Scripts/Inventory/Inventory.cs:85]: Ah je hebt gelijk, mb! Eerste punt staat nog, maar dit hebben we op Discord besproken.

File: aplib.net-demo/Assets/Scripts/Inventory/Key.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    /// <summary>
+    /// the constructor that gives the key its Id
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Inventory/Key.cs:8]: Behavior vs implementation.
```suggestion
    /// Initializes a key with a given ID.
```

File: aplib.net-demo/Assets/Scripts/Inventory/Key.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    /// </summary>
+    /// <param name="constrId">the Id the key is given</param>
+    public Key(int constrId) => id = constrId;
+    private void Start()
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Inventory/Key.cs:12]: Note: I use the new naming of the member `Id` in this suggestion. For future notice; you don't have to call a parameter differently than the member variable. Taking the old code as an example: you can simply change `constrId` to `id` and assign the class member `id` by writing `this.id = id`. The `this` keyword denotes the class we are in.
```suggestion
    /// </summary>
    /// <param name="id">The ID the key is given.</param>
    public Key(int id) => Id = id;
    
    private void Start()
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]

File: aplib.net-demo/Assets/Scripts/Inventory/Key.cs (hunk 4 above)
Comment (wingbladedota) [reply aplib.net-demo/Assets/Scripts/Inventory/Key.cs:13]: Yea, ik weet wel hoe het werkt maar vind dat persoonlijk altijd een beetje verwarrend qua readability - maar heb het aangepast

File: aplib.net-demo/Assets/Scripts/Inventory/Inventory.cs [hunk 5]
```diff
@@ -18,22 +18,19 @@ public class Inventory : MonoBehaviour
...
 
     /// <summary>
-    /// Converts queue to list to check if there are any items with matching names.
-    /// If there are it checks if they are stackable and adds uses. If they are not it does nothing. 
-    /// If there are not matching names it adds the item to the inventory;
+    /// Converts queue to list to check if there are any items with matching names. If there are it checks if they are stackable and adds uses. If they are not it does nothing. If there are not matching names it adds the item to the inventory.
     /// </summary>
-    /// <param name="item">The item that is fed into the inventory.</param>
-    /// <param name="uses">the amount of uses that are added upon pickup.</param>
+    /// <param name="item">The item that is fed into the inventory</param>
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Inventory/Inventory.cs:32]: Nit
```suggestion
    /// <param name="item">The item that is fed into the inventory.</param>
```

File: aplib.net-demo/Assets/Scripts/Inventory/KeyRing.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+using UnityEngine;
+
+public class KeyRing : MonoBehaviour
+{
+    private List<Key> _keyRing;
+
+    /// <summary>
+    /// Creates the keyring
+    /// </summary>
+    private void Start() => _keyRing = new List<Key>();
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Inventory/KeyRing.cs:11]: Also, Awake is the appropriate method for initialization.
```suggestion
```

File: aplib.net-demo/Assets/Scripts/Inventory/KeyRing.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class KeyRing : MonoBehaviour
+{
+    private List<Key> _keyRing;
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Inventory/KeyRing.cs:6]:
```suggestion
    private List<Key> _keyRing = new List<Key>();
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]

File: aplib.net-demo/Assets/Scripts/Doors/Door.cs [hunk 8]
```diff
@@ -28,6 +28,20 @@ private void Awake()
...
+
+        //Once the keys and doors are given matching id's, we can use this
+        /*
+                if (collidingObject.gameObject.CompareTag("Player") && collidingObject.GetComponent<KeyRing>().KeyQuery(this))
+                {
+                    Destroy(transform.parent.gameObject);
+                }*/
     }
+    /// <summary>
+    /// Matches the given key's ID with the door's ID, returns true if the same
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Doors/Door.cs:42]: Nit
```suggestion
    /// Matches the given key's ID with the door's ID, returns true if the same.
```

File: aplib.net-demo/Assets/Scripts/Inventory/Inventory.cs [hunk 9]
```diff
@@ -18,23 +18,20 @@ public class Inventory : MonoBehaviour
...
-    /// Converts queue to list to check if there are any items with matching names.
-    /// If there are it checks if they are stackable and adds uses. If they are not it does nothing. 
-    /// If there are not matching names it adds the item to the inventory;
+    /// Converts queue to list to check if there are any items with matching names. If there are it checks if they are stackable and adds uses. If they are not it does nothing. If there are not matching names it adds the item to the inventory.
     /// </summary>
-    /// <param name="item">The item that is fed into the inventory.</param>
-    /// <param name="uses">the amount of uses that are added upon pickup.</param>
-    public void PickUpItem(Item item, float uses = 1)
+    /// <param name="item">The item that is fed into the inventory</param>
+    /// <param name="uses">The amount of uses that are added upon pickup.</param>
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Inventory/Inventory.cs:33]:
```suggestion
```

File: aplib.net-demo/Assets/Scripts/Inventory/KeyRing.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    /// <returns></returns>
+    public bool KeyQuery(Door door)
+    {
+        foreach (Key k in _keyRing)
+        {
+            _ = door.TryOpenDoor(k);
+        }
+
+        return false;
+
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Inventory/KeyRing.cs:27]: Nit
```suggestion
```

File: aplib.net-demo/Assets/Scripts/Doors/Door.cs [hunk 11]
```diff
@@ -28,6 +28,20 @@ private void Awake()
...
+            Destroy(transform.parent.gameObject);
+
+        //Once the keys and doors are given matching id's, we can use this
+        /*
+                if (collidingObject.gameObject.CompareTag("Player") && collidingObject.GetComponent<KeyRing>().KeyQuery(this))
+                {
+                    Destroy(transform.parent.gameObject);
+                }*/
     }
+    /// <summary>
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Doors/Door.cs:41]: Nit
```suggestion
    }
    
    /// <summary>
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]
Code-wise looks great. Some param summary was left in and an initialization could be done without an Awake/Start method.

**Comment by JensSteenmetz** [comment]
By the way, you can click 'Add suggestion to batch' to commit multiple suggestions at once.

**Review by Tboefijn — CHANGES_REQUESTED** [review]
Left some comments

File: aplib.net-demo/Assets/Scripts/Doors/Door.cs [hunk 12]
```diff
@@ -28,6 +28,20 @@ private void Awake()
...
+    /// Matches the given key's ID with the door's ID, returns true if the same
+    /// </summary>
+    /// <param name="k">The key that is being checked for the same ID as the door</param>
```
Comment (JensSteenmetz) [inline aplib.net-demo/Assets/Scripts/Doors/Door.cs:44]: Nit
```suggestion
    /// <param name="k">The key that is being checked for the same ID as the door.</param>
```

**Review by JensSteenmetz — CHANGES_REQUESTED** [review]

**Review by Tboefijn — APPROVED** [review]
LGTM!

**Pull request merged**